File: scripts/convert_surge_to_shadowrocket.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Surge → Shadowrocket module sync with Rule-Set Expansion
"""

import os
import re
import json
import shutil
import urllib.request
import urllib.error
from pathlib import Path
from datetime import datetime

SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
SURGE_MODULE_DIR = PROJECT_ROOT / "module" / "surge(main)"
SR_MODULE_DIR = PROJECT_ROOT / "module" / "shadowrocket"

# 缓存已下载的规则集
RULESET_CACHE = {}

SURGE_ONLY_KEYS = {
    "use-local-host-item-for-proxy", "encrypted-dns-follow-outbound-mode",
    "encrypted-dns-skip-cert-verification", "force-http-engine-hosts",
    "always-raw-tcp-hosts", "always-raw-tcp-keywords", "tun-included-routes",
}

GENERAL_KEY_MAP = {
    "encrypted-dns-server": "doh-server",
    "tun-excluded-routes":  "bypass-tun",
}

SR_GENERAL_DEFAULTS = """\
bypass-system = true
ipv6 = true
prefer-ipv6 = true
hijack-dns = *:53
dns-direct-fallback-proxy = false"""

RULE_REPLACEMENTS = {
    "REJECT-DROP": "REJECT", "REJECT-TINYGIF": "REJECT", "REJECT-NO-DROP": "REJECT",
}

# Keep remote RULE-SET references for purpose-split AdBlock shards (avoid 50MB+ inline expansion)
PRESERVE_RULESET_MARKERS = (
    "/ruleset/AdBlock/AdBlock_",
    "ruleset%2FAdBlock%2FAdBlock_",
)

REWRITE_MODIFIER_RE = re.compile(r',\s*(extended-matching|pre-matching)\b|'
                                  r'\b(extended-matching|pre-matching)\s*,?')

def fetch_ruleset(url_or_path):
    """抓取并解析规则集"""
    if url_or_path in RULESET_CACHE:
        return RULESET_CACHE[url_or_path]
    
    content = ""
    try:
        # 转换 jsdelivr 链接到本地路径
        if "fastly.jsdelivr.net/gh/nowaytouse/script_hub@master/" in url_or_path:
            rel_path = url_or_path.split("@master/")[-1]
            local_path = PROJECT_ROOT / rel_path
            if local_path.exists():
                print(f"  🏠 Using local file for {url_or_path}")
                content = local_path.read_text(encoding='utf-8')
        
        if not content:
            # 如果是相对路径
            if url_or_path.startswith(".."):
                path = (PROJECT_ROOT / "module/surge(main)/amplify_nexus" / url_or_path).resolve()
                if path.exists():
                    content = path.read_text(encoding='utf-8')
            # 如果是远程URL
            elif url_or_path.startswith("http"):
                print(f"  🌐 Fetching ruleset: {url_or_path}")
                req = urllib.request.Request(url_or_path, headers={'User-Agent': 'ClashMeta/1.18.1'})
                with urllib.request.urlopen(req, timeout=10) as response:
                    content = response.read().decode('utf-8')
        
        # 解析规则
        rules = []
        for line in content.splitlines():
            line = line.strip()
            if not line or line.startswith(("#", "//", ";")):
                continue
            parts = line.split(',')
            rtype = parts[0].upper()
            if rtype in ("DOMAIN", "DOMAIN-SUFFIX", "DOMAIN-KEYWORD", "IP-CIDR", "IP-CIDR6", "GEOIP"):
                rules.append(",".join(parts[:2]))
        
        RULESET_CACHE[url_or_path] = rules
        return rules
    except Exception as e:
        print(f"  ❌ Failed to fetch ruleset {url_or_path}: {e}")
        return []

def convert_content(content: str) -> str:
    lines = content.split('\n')
    out = []
    section = None
    general_defaults_added = False

    for line in lines:
        stripped = line.strip()

        if stripped.startswith('[') and stripped.endswith(']'):
            section = stripped[1:-1]
            out.append(line)
            if section == "General" and not general_defaults_added:
                out.append(SR_GENERAL_DEFAULTS)
                general_defaults_added = True
            continue

        if re.match(r'^#!', line):
            if re.match(r'^#!\s*(update-interval|ability)\s*=', line, re.IGNORECASE):
                continue
            m = re.match(r'^#!\s*(\S+?)\s*=\s*(.*)$', line)
            if m:
                key, val = m.group(1).strip(), m.group(2)
                if key == "desc":
                    val = f"[🚀SR] {val}" if "[🚀SR]" not in val else val
                out.append(f"#!{key}={val}")
            continue

        if section == "General" and not stripped.startswith('#') and stripped:
            if any(k in line for k in SURGE_ONLY_KEYS):
                out.append(f"# [SR不支持] {line.lstrip()}")
                continue
            for k, v in GENERAL_KEY_MAP.items():
                if line.startswith(k):
                    line = line.replace(k, v)
            out.append(line)
            continue

        if section == "Rule" and not stripped.startswith('#') and stripped:
            if stripped.startswith('RULE-SET,'):
                parts = stripped.split(',')
                url_or_path = parts[1].strip()
                policy = parts[2].strip() if len(parts) > 2 else "REJECT"
                policy = RULE_REPLACEMENTS.get(policy, policy)

                if any(marker in url_or_path for marker in PRESERVE_RULESET_MARKERS):
                    cleaned = REWRITE_MODIFIER_RE.sub('', line)
                    cleaned = re.sub(r',update-interval=\d+', '', cleaned)
                    cleaned = re.sub(r',no-resolve', '', cleaned)
                    for old, new in RULE_REPLACEMENTS.items():
                        cleaned = cleaned.replace(old, new)
                    out.append(cleaned.strip())
                    continue

                print(f"  📦 Expanding RULE-SET: {url_or_path}")
                expanded_rules = fetch_ruleset(url_or_path)
                if expanded_rules:
                    out.append(f"# --- Expanded from {url_or_path} ---")
                    for r in expanded_rules:
                        out.append(f"{r},{policy}")
                    out.append(f"# --- End expansion ---")
                else:
                    out.append(f"# [无法展开] {line}")
                continue
            
            if stripped.startswith('PROTOCOL,'):
                out.append(f"# [SR不支持PROTOCOL] {line}")
                continue

        line = re.sub(r'%(?:INSERT|APPEND)%\s*', '', line)
        for old, new in RULE_REPLACEMENTS.items():
            line = line.replace(old, new)
        line = REWRITE_MODIFIER_RE.sub('', line)
        line = re.sub(r',"update-interval=\d+"', '', line)
        
        out.append(line)

    return '\n'.join(out)

def process_all_modules():
    if not SR_MODULE_DIR.exists():
        SR_MODULE_DIR.mkdir(parents=True)

    categories = ["amplify_nexus", "head_expanse", "narrow_pierce"]
    stats = {"total": 0, "converted": 0, "failed": 0}
    
    for cat in categories:
        (SR_MODULE_DIR / cat).mkdir(exist_ok=True)
        cat_path = SURGE_MODULE_DIR / cat
        if not cat_path.exists(): continue

        for module_file in sorted(cat_path.glob("*.sgmodule")):
            stats["total"] += 1
            print(f"🔄 Converting: {module_file.name}")
            try:
                content = module_file.read_text(encoding="utf-8")
                converted = convert_content(content)
                out_path = SR_MODULE_DIR / cat / (module_file.stem + ".module")
                out_path.write_text(converted, encoding="utf-8")
                stats["converted"] += 1
            except Exception as e:
                print(f"  ❌ Failed {module_file.name}: {e}")
                stats["failed"] += 1

    return stats

if __name__ == "__main__":
    s = process_all_modules()
    print(f"\n✅ All modules converted: {s['converted']}/{s['total']}")

File: create/tools/convert_surge_to_shadowrocket.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/nowaytouse/script_hub/create/hub"
)

var (
	surgeOnlyKeys = map[string]bool{
		"use-local-host-item-for-proxy": true, "encrypted-dns-follow-outbound-mode": true,
		"encrypted-dns-skip-cert-verification": true, "force-http-engine-hosts": true,
		"always-raw-tcp-hosts": true, "always-raw-tcp-keywords": true, "tun-included-routes": true,
		"compatibility-mode": true, "http-api-tls": true, "http-api-web-dashboard": true,
		"allow-wifi-access": true, "allow-hotspot-access": true, "wifi-access-http-port": true,
		"wifi-access-socks5-port": true, "proxy-restricted-to-lan": true, "block-quic": true,
		"exclude-simple-hostnames": true, "read-etc-hosts": true, "include-apns": true,
		"include-cellular-services": true, "wifi-assist": true, "allow-dns-svcb": true,
		"ipv6-vif": true, "include-all-networks": true, "include-local-networks": true,
		"auto-suspend": true, "all-hybrid": true, "http-listen": true, "socks5-listen": true,
		"proxy-test-udp": true,
	}

	generalKeyMap = map[string]string{
		"encrypted-dns-server": "dns-server",
		"dns-server":           "fallback-dns-server",
		"tun-excluded-routes":  "tun-excluded-routes",
	}

	ruleReplacements = map[string]string{
		"REJECT-DROP": "REJECT", "REJECT-TINYGIF": "REJECT", "REJECT-NO-DROP": "REJECT",
	}

	preserveRulesetMarkers = []string{
		"/rulesets/AdBlock/AdBlock_",
		"ruleset%2FAdBlock%2FAdBlock_",
		"skk_upstream/",
		"HTTPDNS_Hijack.list",
		"nowaytouse/script_hub@master/rulesets/",
	}

	rewriteModifierRe = regexp.MustCompile(`,\s*(extended-matching|pre-matching)\b|\b(extended-matching|pre-matching)\s*,?`)

	srGeneralDefaults = `bypass-system = true
ipv6 = true
prefer-ipv6 = true
hijack-dns = *:53
dns-direct-fallback-proxy = false`

	hostKeepPatterns = []string{
		`^dns\.`, `^doh\.`, `^cloudflare`, `\.google\.com$`, `^talk\.google`, `^mtalk\.google`,
		`^alt\d+-mtalk`, `^stun`, `^connectivitycheck`, `^detectportal`, `^msftconnecttest`, `^msftncsi`,
		`^\*\.cn$`, `^\*\.tw$`, `^\*\.cht\.com\.tw$`, `^\*\.hinet\.net$`, `^\*\.he\.net$`,
		`^hanime1\.me$`, `^18comic\.vip$`, `^3hentai\.net$`, `^router\.`, `^miwifi\.`, `^tplogin\.`,
		`\.liangxin1\.xyz$`, `^raw\.githubusercontent`, `^github\.com$`, `^\d+\.\d+\.\d+\.\d+$`,
	}

	promaxSurgeModules = []string{
		"🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).sgmodule",
		"📱 Universal Ad-Blocking Rules (PROMAX Lite).sgmodule",
	}

	rulesetCache = make(map[string][]string)
)

func shouldKeepHostLine(line string) bool {
	stripped := strings.TrimSpace(line)
	if stripped == "" || strings.HasPrefix(stripped, "#") {
		return true
	}
	if strings.Contains(stripped, "=") {
		domain := strings.TrimSpace(strings.Split(stripped, "=")[0])
		for _, pattern := range hostKeepPatterns {
			re := regexp.MustCompile("(?i)" + pattern)
			if re.MatchString(domain) {
				return true
			}
		}
	}
	return false
}

func fetchRuleset(urlOrPath string) []string {
	if cached, ok := rulesetCache[urlOrPath]; ok {
		return cached
	}
	if len(rulesetCache) >= 100 {
		rulesetCache = make(map[string][]string)
	}

	var content string

	cdnPatterns := []string{
		"fastly.jsdelivr.net/gh/nowaytouse/script_hub@master/",
		"cdn.jsdelivr.net/gh/nowaytouse/script_hub@master/",
		"gcore.jsdelivr.net/gh/nowaytouse/script_hub@master/",
	}

	for _, cdn := range cdnPatterns {
		if strings.Contains(urlOrPath, cdn) {
			parts := strings.Split(urlOrPath, "@master/")
			if len(parts) == 2 {
				localPath := filepath.Join(hub.ROOT, parts[1])
				if hub.ValidateFileExists(localPath, "") {
					content = hub.ReadFileString(localPath)
					break
				}
			}
		}
	}

	if content == "" && strings.HasPrefix(urlOrPath, "..") {
		localPath := filepath.Join(hub.ROOT, "modules/surge/amplify_nexus", urlOrPath)
		if hub.ValidateFileExists(localPath, "") {
			content = hub.ReadFileString(localPath)
		}
	}

	if content == "" && strings.HasPrefix(urlOrPath, "http") {
		// Download with retries
		content = hub.SafeDownload(urlOrPath, 3, 10)
	}

	if content == "" {
		return nil
	}

	var rules []string
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(line, "//") || strings.HasPrefix(line, ";") {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) == 0 {
			continue
		}
		rtype := strings.ToUpper(parts[0])
		if rtype == "DOMAIN" || rtype == "DOMAIN-SUFFIX" || rtype == "DOMAIN-KEYWORD" || rtype == "IP-CIDR" || rtype == "IP-CIDR6" || rtype == "GEOIP" {
			if len(parts) >= 2 {
				rules = append(rules, parts[0]+","+parts[1])
			}
		}
	}

	rulesetCache[urlOrPath] = rules
	return rules
}

func convertContent(content string, moduleStem string) string {
	if moduleStem == "" {
		meta, _ := hub.ParseModule(content)
		moduleStem = hub.ModuleStemFromMeta(meta, "")
	}

	lines := strings.Split(content, "\n")
	var out []string
	var section string
	generalDefaultsAdded := false

	for _, line := range lines {
		stripped := strings.TrimSpace(line)

		if strings.HasPrefix(stripped, "[") && strings.HasSuffix(stripped, "]") {
			section = stripped[1 : len(stripped)-1]
			out = append(out, line)
			if section == "General" && !generalDefaultsAdded {
				out = append(out, srGeneralDefaults)
				generalDefaultsAdded = true
			}
			continue
		}

		if strings.HasPrefix(line, "#!") {
			if regexp.MustCompile(`(?i)^#!\s*(update-interval|ability)\s*=`).MatchString(line) {
				continue
			}
			m := regexp.MustCompile(`^#!\s*(\S+?)\s*=\s*(.*)$`).FindStringSubmatch(line)
			if m != nil {
				key, val := strings.TrimSpace(m[1]), m[2]
				if key == "desc" && !strings.Contains(val, "[🚀SR]") {
					val = "[🚀SR] " + val
				}
				out = append(out, fmt.Sprintf("#!%s=%s", key, val))
			} else {
				out = append(out, line)
			}
			continue
		}

		if section == "General" && !strings.HasPrefix(stripped, "#") && stripped != "" {
			hasSurgeKey := false
			for k := range surgeOnlyKeys {
				if strings.Contains(line, k) {
					hasSurgeKey = true
					break
				}
			}
			if hasSurgeKey {
				out = append(out, "# [SR does not support ] "+strings.TrimLeft(line, " \t"))
				continue
			}
			for k, v := range generalKeyMap {
				if strings.HasPrefix(line, k) {
					line = strings.Replace(line, k, v, 1)
				}
			}
			out = append(out, line)
			continue
		}

		if section == "Rule" && !strings.HasPrefix(stripped, "#") && stripped != "" {
			if strings.HasPrefix(stripped, "RULE-SET,") {
				parts := strings.Split(stripped, ",")
				urlOrPath := strings.TrimSpace(parts[1])
				policy := "REJECT"
				if len(parts) > 2 {
					policy = strings.TrimSpace(parts[2])
				}
				if repl, ok := ruleReplacements[policy]; ok {
					policy = repl
				}

				preserve := false
				for _, marker := range preserveRulesetMarkers {
					if strings.Contains(urlOrPath, marker) {
						preserve = true
						break
					}
				}

				if preserve {
					cleaned := rewriteModifierRe.ReplaceAllString(line, "")
					cleaned = regexp.MustCompile(`,update-interval=\d+`).ReplaceAllString(cleaned, "")
					cleaned = regexp.MustCompile(`,no-resolve`).ReplaceAllString(cleaned, "")
					for old, new := range ruleReplacements {
						cleaned = strings.ReplaceAll(cleaned, old, new)
					}
					out = append(out, strings.TrimSpace(cleaned))
					continue
				}

				fmt.Printf("  📦 Expanding RULE-SET: %s\n", urlOrPath)
				expandedRules := fetchRuleset(urlOrPath)
				if len(expandedRules) > 0 {
					out = append(out, fmt.Sprintf("# --- Expanded from %s ---", urlOrPath))
					for _, r := range expandedRules {
						out = append(out, fmt.Sprintf("%s,%s", r, policy))
					}
					out = append(out, "# --- End expansion ---")
				} else {
					out = append(out, fmt.Sprintf("# [Cannot expand] %s", line))
				}
				continue
			}

			if strings.HasPrefix(stripped, "PROTOCOL,") {
				out = append(out, "# [SR does not support PROTOCOL] "+line)
				continue
			}
		}

		if section == "Script" && stripped != "" && !strings.HasPrefix(stripped, "#") {
			line = hub.AdaptScriptLineForSr(line, moduleStem)
		} else if section == "MITM" && stripped != "" && !strings.HasPrefix(stripped, "#") {
			line = hub.AdaptMitmLineForSr(line)
		}

		if section != "MITM" {
			line = regexp.MustCompile(`%(?:INSERT|APPEND)%\s*`).ReplaceAllString(line, "")
		}
		for old, new := range ruleReplacements {
			line = strings.ReplaceAll(line, old, new)
		}
		line = rewriteModifierRe.ReplaceAllString(line, "")
		line = regexp.MustCompile(`,"update-interval=\d+"`).ReplaceAllString(line, "")

		out = append(out, line)
	}

	return hub.SanitizeFileContent(strings.Join(out, "\n"), true)
}

func convertDevtoolsModuleForSr(content string) string {
	meta, sections := hub.ParseModule(content)
	moduleStem := hub.ModuleStemFromMeta(meta, hub.DEVTOOLS_STEM)
	converted := convertContent(content, moduleStem)
	meta, sections = hub.ParseModule(converted)
	desc := meta["desc"]
	if !strings.Contains(desc, "[🚀SR]") {
		meta["desc"] = "[🚀SR] " + desc + `\nSub-Store simplified for Surge-Noability; produce cron only available on Surge`
	}
	header := hub.FormatHeader(meta, nil)
	ordered := make([]hub.ModuleSection, 0)
	for _, s := range sections {
		ordered = append(ordered, hub.ModuleSection{Name: s.Name, Lines: s.Lines})
	}
	return hub.SanitizeFileContent(hub.FormatModule(header, ordered, true, meta), true)
}

func ConvertPromaxModules() bool {
	ok := true
	subdirs := []string{"", "github"}
	for _, subdir := range subdirs {
		srcCat := filepath.Join(hub.ROOT, "modules/surge/head_expanse")
		outCat := filepath.Join(hub.ROOT, "modules/shadowrocket/head_expanse")
		if subdir != "" {
			srcCat = filepath.Join(srcCat, subdir)
			outCat = filepath.Join(outCat, subdir)
		}
		hub.EnsureDir(outCat)

		for _, name := range promaxSurgeModules {
			src := filepath.Join(srcCat, name)
			if !hub.ValidateFileExists(src, "") {
				if subdir == "" {
					fmt.Printf("⚠️ PROMAX source missing: %s\n", name)
					ok = false
				}
				continue
			}
			subdirName := subdir
			if subdirName == "" {
				subdirName = "cdn"
			}
			fmt.Printf("Converting PROMAX → SR [%s]: %s\n", subdirName, name)

			content := hub.ReadFileString(src)
			converted := convertContent(content, "")

			outPath := filepath.Join(outCat, strings.TrimSuffix(name, ".sgmodule")+".module")
			hub.SafeWriteFile(outPath, converted, true)
			fmt.Printf("✅ SR module written: %s\n", outPath)
		}
	}
	return ok
}

func ProcessAllModules() (int, int, int) {
	srModuleDir := filepath.Join(hub.ROOT, "modules/shadowrocket")
	surgeModuleDir := filepath.Join(hub.ROOT, "modules/surge")
	hub.EnsureDir(srModuleDir)

	categories := []string{"amplify_nexus", "head_expanse"}
	total, converted, failed := 0, 0, 0

	for _, cat := range categories {
		hub.EnsureDir(filepath.Join(srModuleDir, cat))
		catPath := filepath.Join(surgeModuleDir, cat)
		if !hub.ValidateDirExists(catPath, "") {
			continue
		}

		entries, _ := os.ReadDir(catPath)
		var files []string
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".sgmodule") {
				files = append(files, e.Name())
			}
		}
		sort.Strings(files)

		for _, file := range files {
			total++
			fmt.Printf("🔄 Converting: %s\n", file)

			moduleFile := filepath.Join(catPath, file)
			content := hub.ReadFileString(moduleFile)

			stem := strings.TrimSuffix(file, ".sgmodule")
			var convertedStr string
			if stem == hub.DEVTOOLS_STEM {
				convertedStr = convertDevtoolsModuleForSr(content)
			} else {
				convertedStr = convertContent(content, stem)
			}

			outPath := filepath.Join(srModuleDir, cat, stem+".module")
			hub.SafeWriteFile(outPath, convertedStr, true)
			converted++
		}
	}
	return total, converted, failed
}

func convertProxyGroupLine(line string) string {
	line = regexp.MustCompile(`,\s*no-alert=\d+`).ReplaceAllString(line, "")
	line = regexp.MustCompile(`,\s*hidden=\d+`).ReplaceAllString(line, "")
	line = regexp.MustCompile(`,\s*persistent=\d+`).ReplaceAllString(line, "")
	line = regexp.MustCompile(`,\s*evaluate-before-use=\d+`).ReplaceAllString(line, "")
	line = regexp.MustCompile(`,\s*include-all-proxies=\d+`).ReplaceAllString(line, "")
	line = regexp.MustCompile(`,\s*icon-url=[^\s,]+`).ReplaceAllString(line, "")

	line = regexp.MustCompile(`=\s*smart,`).ReplaceAllString(line, "= url-test,")

	if strings.Contains(line, "include-other-group=") && !strings.Contains(line, "policy-path=") {
		match := regexp.MustCompile(`include-other-group="([^"]+)"`).FindStringSubmatch(line)
		if match != nil {
			line = regexp.MustCompile(`include-other-group="[^"]+"`).ReplaceAllString(line, "")
		}
	}
	return line
}

func convertMainConfig(surgeConfPath string, outputPath string, compactHost bool) bool {
	fmt.Printf("Converting main config: %s\n", filepath.Base(surgeConfPath))

	if !hub.ValidateFileExists(surgeConfPath, "") {
		fmt.Printf("Config file not found: %s\n", surgeConfPath)
		return false
	}

	content := hub.ReadFileString(surgeConfPath)
	lines := strings.Split(content, "\n")
	var out []string
	var section string
	hostLinesKept := 0
	hostLinesSkipped := 0

	var dnsServerLine string
	var fallbackDnsLine string

	for _, line := range lines {
		stripped := strings.TrimSpace(line)

		if strings.HasPrefix(stripped, "[") && strings.HasSuffix(stripped, "]") {
			if section == "General" && (dnsServerLine != "" || fallbackDnsLine != "") {
				if dnsServerLine != "" {
					out = append(out, dnsServerLine)
				}
				if fallbackDnsLine != "" {
					out = append(out, fallbackDnsLine)
				}
				dnsServerLine = ""
				fallbackDnsLine = ""
			}

			section = stripped[1 : len(stripped)-1]
			out = append(out, line)

			if section == "Host" && compactHost {
				out = append(out, "# Shadowrocket Compact - Keep only critical DNS config")
				out = append(out, "# Full domain-level DoH split via [General] doh-server global config")
			}
			continue
		}

		if section == "Ponte" {
			if !strings.HasPrefix(stripped, "#") && stripped != "" {
				out = append(out, "# [SR does not support Ponte] "+line)
			} else {
				out = append(out, line)
			}
			continue
		}

		if section == "General" && !strings.HasPrefix(stripped, "#") && stripped != "" {
			hasSurgeKey := false
			for k := range surgeOnlyKeys {
				if strings.Contains(line, k) {
					hasSurgeKey = true
					break
				}
			}
			if hasSurgeKey {
				out = append(out, "# [SR does not support ] "+strings.TrimLeft(line, " \t"))
				continue
			}

			converted := false
			foundK := ""
			for k, v := range generalKeyMap {
				if strings.HasPrefix(line, k) {
					line = strings.Replace(line, k, v, 1)
					converted = true
					if k == "encrypted-dns-server" {
						dnsServerLine = line
						foundK = k
					} else if k == "dns-server" {
						fallbackDnsLine = line
						foundK = k
					}
					break
				}
			}

			if !converted || (foundK != "encrypted-dns-server" && foundK != "dns-server") {
				out = append(out, line)
			}
			continue
		}

		if section == "Proxy Group" && !strings.HasPrefix(stripped, "#") && stripped != "" && strings.Contains(stripped, "=") {
			line = convertProxyGroupLine(line)
			out = append(out, line)
			continue
		}

		if section == "Rule" && !strings.HasPrefix(stripped, "#") && stripped != "" {
			line = regexp.MustCompile(`,\s*extended-matching`).ReplaceAllString(line, "")
			line = regexp.MustCompile(`,\s*no-resolve\s*$`).ReplaceAllString(line, "")
			if strings.HasPrefix(stripped, "PROTOCOL,") {
				out = append(out, "# [SR does not support PROTOCOL] "+line)
				continue
			}
			out = append(out, line)
			continue
		}

		if section == "Host" {
			if compactHost {
				if shouldKeepHostLine(line) {
					line = strings.ReplaceAll(line, "server:force-syslib", "server:syslib")
					out = append(out, line)
					if !strings.HasPrefix(stripped, "#") && stripped != "" {
						hostLinesKept++
					}
				} else {
					hostLinesSkipped++
				}
			} else {
				line = strings.ReplaceAll(line, "server:force-syslib", "server:syslib")
				out = append(out, line)
			}
			continue
		}

		if section == "Script" {
			if !strings.HasPrefix(stripped, "#") && stripped != "" {
				out = append(out, "# [SR does not support Script] "+line)
			} else {
				out = append(out, line)
			}
			continue
		}

		out = append(out, line)
	}

	result := strings.Join(out, "\n")
	hub.SafeWriteFile(outputPath, result, true)
	fmt.Printf("Config converted: %s\n", outputPath)
	if compactHost {
		fmt.Printf("Host compact: Kept %d lines, skipped %d lines\n", hostLinesKept, hostLinesSkipped)
	}
	fmt.Printf("总lines数: %d\n", len(out))
	return true
}

// RunConvertSurgeToShadowrocket takes an args map. It supports modules, promaxOnly, config, output, fullHost.
func RunConvertSurgeToShadowrocket(args map[string]string) int {
	_, modules := args["modules"]
	_, promaxOnly := args["promax-only"]
	config := args["config"]
	output := args["output"]
	_, fullHost := args["full-host"]

	if config != "" {
		if !hub.ValidateFileExists(config, "") {
			fmt.Printf("Config file not found: %s\n", config)
			return 1
		}
		if output == "" {
			output = filepath.Join(filepath.Dir(config), strings.TrimSuffix(filepath.Base(config), filepath.Ext(config))+"_Shadowrocket.conf")
		}
		success := convertMainConfig(config, output, !fullHost)
		if success {
			return 0
		}
		return 1
	}

	if promaxOnly {
		if ConvertPromaxModules() {
			return 0
		}
		return 1
	}

	if modules {
		total, converted, failed := ProcessAllModules()
		fmt.Printf("All modules converted: %d/%d\n", converted, total)
		if failed > 0 {
			fmt.Printf("Failed: %d modules\n", failed)
		}
		if converted == 0 {
			return 1
		}
		return 0
	}

	surgeConf := filepath.Join(hub.ROOT, ".claude", "NyaMiiKo.conf.conf")
	if hub.ValidateFileExists(surgeConf, "") {
		output := filepath.Join(hub.ROOT, ".claude", "NyaMiiKo_Shadowrocket.conf")
		fmt.Println("Converting default project config")
		success := convertMainConfig(surgeConf, output, !fullHost)
		if success {
			return 0
		}
		return 1
	}

	fmt.Println("Not found .claude/NyaMiiKo.conf.conf, skipping main config, converting modules only")
	total, converted, failed := ProcessAllModules()
	fmt.Printf("All modules converted: %d/%d\n", converted, total)
	if failed > 0 {
		fmt.Printf("Failed: %d modules\n", failed)
	}
	if converted == 0 {
		return 1
	}
	return 0
}
